lib/segment/src/index/field_index/full_text_index/posting_list.rs

```diff
@@ -37,23 +40,305 @@ impl PostingList {
         self.list.binary_search(val).is_ok()
     }
 
-    pub fn iter<'a>(
-        &'a self,
-        filter: impl Fn(PointOffsetType) -> bool + 'a,
-    ) -> impl Iterator<Item = PointOffsetType> + 'a {
-        self.list.iter().copied().filter(move |&idx| filter(idx))
+    pub fn iter(&self) -> impl Iterator<Item = PointOffsetType> + '_ {
+        self.list.iter().copied()
     }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CompressedPostingList {
+    len: usize,
+    last_doc_id: PointOffsetType,
+    data: Vec<u8>,
+    chunks: Vec<CompressedPostingChunk>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CompressedPostingChunk {
+    initial: PointOffsetType,
+    offset: u32,
+}
+
+impl CompressedPostingList {
+    pub fn new(mut posting_list: PostingList) -> Self {
+        if posting_list.list.is_empty() {
+            return Self::default();
+        }
+        let len = posting_list.len();
+        let last_doc_id = *posting_list.list.last().unwrap();
 
-    pub fn shrink_to_fit(&mut self) {
-        self.list.shrink_to_fit();
+        let bitpacker = BitPackerImpl::new();
+        posting_list.list.sort_unstable();
+
+        let last = *posting_list.list.last().unwrap();
+        while posting_list.list.len() % BitPackerImpl::BLOCK_LEN != 0 {
+            posting_list.list.push(last);
```

Comment:
I am not sure this is a good heuristic.
If we assume that data is distributed according to Zipf, only first few thousands of the vocab tokens will have enough occurances, while the long tail of the smaller posting lists will actually be bloated with this extra "alignment".

I propose to make a Enum of postings, which can deside for it's own which impelemtation to choose.

Reply:
Fixed. I implemented it simpler. Remainder data is uncompressed. For small postings list we contain add data as remainder